### PR TITLE
Add libcudf wrappers around current_device_resource functions.

### LINKED ISSUE #16676: [FEA] Add libcudf cudf::get_current_device_resource() wrapper for rmm::get_current_device_resource()
**Is your feature request related to a problem? Please describe.**
I'm about to undertake yet another change to every function in libcudf that takes an MR parameter. To prevent doing this too many times in the future, I would like to provide a wrapper in libcudf to centralize calls to `rmm::get_current_device_resource[_ref]()` and an alias for RMM `resource_ref` type. hopefully this can reduce the number of files that need to change in the future when this functionality changes (e.g. if it moves out of RMM to CCCL).

**Describe the solution you'd like**
A header, say `cpp/include/cudf/utilities/memory_resource.hpp`. Please provide suggestions on the best place for this. Contents something like this:

```
namespace cudf {

using device_async_resource_ref = rmm::device_async_resource_ref;

inline device_async_resource_ref get_current_device_resource_ref() {
  return rmm::get_current_device_resource_ref();
}

inline device_async_resource_ref set_current_device_resource_ref(device_async_resource_ref mr) {
  return rmm::set_current_device_resource_ref(mr);
}

inline device_async_resource_ref reset_current_device_resource_ref() {
  return rmm::reset_current_device_resource_ref();
}
```

**Describe alternatives you've considered**

**Additional context**
https://github.com/rapidsai/rmm/pull/1598

### PR DESCRIPTION
**Pull request by harrism**
## Description
Merge after rapidsai/rmm#1661

Creates and uses CUDF internal wrappers around RMM `current_device_resource` functions.

I've marked this PR as breaking because it breaks the ABI, however the API is compatible.

For reviewers, the most substantial additions are in the new file `<cudf/utilities/memory_resource.hpp>`, and in the `DEVELOPER_GUIDE.md` and `*.rst` docs. The rest are all replacements of an include and all calls to `rmm::get_current_device_resource()` with `cudf::get_current_device_resource_ref()`.

Closes #16676

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by harrism** [comment]
@vyasr I could use your help understanding the docs build failure. Thanks!

**Comment by vyasr** [comment]
You'll need to add a new docs page for memory_resource.hpp to https://github.com/rapidsai/cudf/tree/branch-24.10/docs/cudf/source/libcudf_docs/api_docs, and then add it either to [index.rst](https://github.com/rapidsai/cudf/blob/branch-24.10/docs/cudf/source/libcudf_docs/api_docs/index.rst) or to one of the subpages that it lists in its ToC (or their subpages, etc).

**Comment by harrism** [comment]
> You'll need to add a new docs page for memory_resource.hpp to https://github.com/rapidsai/cudf/tree/branch-24.10/docs/cudf/source/libcudf_docs/api_docs, and then add it either to [index.rst](https://github.com/rapidsai/cudf/blob/branch-24.10/docs/cudf/source/libcudf_docs/api_docs/index.rst) or to one of the subpages that it lists in its ToC (or their subpages, etc).

Thanks! Made an attempt.

**Review by ttnghia — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

**Comment by nvdbaranec** [comment]
This may have been asked already, but why do we have to include `<cudf/utilities/default_stream.hpp>` as well as `<cudf/utilities/memory_resource.hpp>` ?

**Review by nvdbaranec — COMMENTED** [review]
There are a handful of places where the default stream header is included by ultimately unused. Example:

https://github.com/rapidsai/cudf/blob/f83b64dcca7ba898613b5e4f0e6a49283467a473/cpp/include/cudf/reduction/detail/segmented_reduction_functions.hpp#L22

**Comment by harrism** [comment]
> This may have been asked already, but why do we have to include `<cudf/utilities/default_stream.hpp>` as well as `<cudf/utilities/memory_resource.hpp>` ? Or were the few places where this was done because it wasn't explicitly included for the neighboring `cudf::get_default_stream`?

I don't understand the question. One header has the default stream functions, the other has the current device resource functions. They are different.

**Review by nvdbaranec — APPROVED** [review]
Just that one lingering unnecessary header - not a blocker.

File: cpp/include/cudf/io/text/detail/tile_state.hpp [hunk 1]
```diff
@@ -17,8 +17,7 @@
 #pragma once
 
 #include <cudf/utilities/export.hpp>
-
-#include <rmm/resource_ref.hpp>
+#include <cudf/utilities/memory_resource.hpp>
```
Comment (davidwendt) [inline cpp/include/cudf/io/text/detail/tile_state.hpp:20]: Loosk like this include can be removed.

File: cpp/include/cudf/io/text/detail/trie.hpp [hunk 2]
```diff
@@ -19,11 +19,11 @@
 #include <cudf/detail/utilities/vector_factories.hpp>
 #include <cudf/io/text/detail/multistate.hpp>
 #include <cudf/utilities/export.hpp>
+#include <cudf/utilities/memory_resource.hpp>
```
Comment (davidwendt) [inline cpp/include/cudf/io/text/detail/trie.hpp:22]: Looks like this include can be removed.

**Review by davidwendt — APPROVED** [review]

**Pull request merged**